### PR TITLE
Feat/facet attribute

### PR DESCRIPTION
**Pull request by panquez**

### CODE REVIEW

File: include/geode/mesh/core/detail/vertex_cycle.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                auto other_min_itr = std::min_element(
+                    other_vertices.begin(), other_vertices.end() );
+                if( *min_itr != *other_min_itr )
+                {
+                    return false;
+                }
+                std::rotate( other_vertices.begin(), other_min_itr,
+                    other_vertices.end() );
+                std::rotate(
+                    vertices_copy.begin(), min_itr, vertices_copy.end() );
```
Comment (BotellaA) [inline include/geode/mesh/core/detail/vertex_cycle.h]: Why not doing this once in the constructor?

File: src/geode/mesh/builder/polygonal_surface_builder.cpp [hunk 2]
```diff
@@ -278,13 +291,47 @@ namespace geode
...
+    template < index_t dimension >
+    void PolygonalSurfaceBuilder< dimension >::update_edge_vertices(
+        const std::vector< index_t >& old2new )
+    {
+        polygonal_surface_.update_edge_vertices( old2new );
+    }
+
+    template < index_t dimension >
+    void PolygonalSurfaceBuilder< dimension >::delete_edges(
+        const std::vector< index_t >& old2new )
```
Comment (BotellaA) [inline src/geode/mesh/builder/polygonal_surface_builder.cpp]: API seems strange, "delete_edges" and we have a mapping as parameter

File: src/geode/mesh/core/polygonal_surface.cpp [hunk 3]
```diff
@@ -121,11 +122,81 @@ namespace geode
...
+                return edge_indices_.at( edge_vertices );
+            }
+            return NO_ID;
+        }
+
+        index_t find_or_create_edge(
+            const std::array< index_t, 2 >& edge_vertices )
+        {
+            auto size = edge_indices_.size();
+            std::vector< index_t > vertices( 2 );
```
Comment (BotellaA) [inline src/geode/mesh/core/polygonal_surface.cpp]:
```suggestion
            std::vector< index_t > vertices{ edge_vertices[0], edge_vertices[1] };
```
You choose

File: src/geode/mesh/core/polygonal_surface.cpp [hunk 4]
```diff
@@ -121,11 +122,81 @@ namespace geode
...
+            }
+            edge_indices_[vertices] = size;
+            edge_attribute_manager_.resize( size + 1 );
+            return size;
+        }
+
+        void update_edge_vertices( const std::vector< index_t >& old2new )
+        {
+            auto old_edge_indices = std::move( edge_indices_ );
+            edge_indices_.clear();
```
Comment (BotellaA) [inline src/geode/mesh/core/polygonal_surface.cpp]: the use of a variable after a move is undefined behavior. Don't do this! You are lucky it works.

File: src/geode/mesh/core/polygonal_surface.cpp [hunk 5]
```diff
@@ -121,11 +122,81 @@ namespace geode
...
+                detail::VertexCycle updated_cycle{ updated_vertices };
+                edge_indices_[updated_cycle] = cycle.second;
+            }
+        }
+
+        void delete_edges( const std::vector< index_t >& old2new )
+        {
+            std::vector< detail::VertexCycle > key_to_erase;
+            key_to_erase.reserve( old2new.size() );
+            for( auto cycle : edge_indices_ )
```
Comment (BotellaA) [inline src/geode/mesh/core/polygonal_surface.cpp]: you seems to copy every edge_indices_ entry. Not sure if this is the best idea.

File: src/geode/mesh/core/polyhedral_solid.cpp [hunk 6]
```diff
@@ -108,11 +109,78 @@ namespace geode
...
+                return id;
+            }
+            facet_indices_[facet_vertices] = size;
+            facet_attribute_manager_.resize( size + 1 );
+            return size;
+        }
+
+        void update_facet_vertices( const std::vector< index_t >& old2new )
+        {
+            auto old_facet_indices = std::move( facet_indices_ );
```
Comment (BotellaA) [inline src/geode/mesh/core/polyhedral_solid.cpp]: Same usage of the std::move

File: src/geode/mesh/core/polyhedral_solid.cpp [hunk 7]
```diff
@@ -108,11 +109,78 @@ namespace geode
...
+                detail::VertexCycle updated_cycle{ updated_vertices };
+                facet_indices_[updated_cycle] = cycle.second;
+            }
+        }
+
+        void delete_facets( const std::vector< index_t >& old2new )
+        {
+            std::vector< detail::VertexCycle > key_to_erase;
+            key_to_erase.reserve( old2new.size() );
+            for( auto cycle : facet_indices_ )
```
Comment (BotellaA) [inline src/geode/mesh/core/polyhedral_solid.cpp]: Same question for the copy

File: include/geode/mesh/core/detail/vertex_cycle.h (hunk 1 above)
Comment (panquez) [reply include/geode/mesh/core/detail/vertex_cycle.h]: Good idea! Done!

**Pull request merged**